coref/models/general_coref_model.py

```diff
@@ -320,28 +324,35 @@ def run(
         # cluster_ids     [n_words]
         words, cluster_ids = self.we(doc, encoded_doc)
 
-        # Obtain bilinear scores and leave only top-k antecedents for each word
-        # top_rough_scores  [n_words, n_ants]
-        # top_indices       [n_words, n_ants]
-        top_rough_scores, top_indices = cast(
-            Tuple[torch.Tensor, torch.Tensor], self.rough_scorer(words)
+        # If scoring_fn is None
+        #     Obtain bilinear scores and leave only top-k antecedents for each word
+        #     rough_scores         [n_words, n_ants]
+        #     rough_scores_indices [n_words, n_ants]
+        # Else
+        #     sorted scores given scoring fn and their original  positions
+        #     rough_scores         [n_words, 1]
+        #     rough_scores_indices [n_words, 1]
+        rough_scores, rough_scores_indices = cast(
+            Tuple[torch.Tensor, torch.Tensor],
+            self.rough_scorer(words, scoring_fn),
         )
         if return_mention:
-            return doc.subwords_2_words(
-                set(int(i) for i in top_indices.flatten().tolist())
+            return doc.subwords_2_words_w_payload(
+                [int(i) for i in rough_scores_indices.flatten().tolist()],
+                rough_scores.flatten().tolist(),
```

Comment:
How does this payload affect the mapping?

Reply:
It the payload is the criterium of sorting in the token selection step